### PR TITLE
v6 actors upgrade

### LINKED ISSUE #1255: implement FIP-0020: Add return value to WithdrawBalance
**Task summary**

- Add return value to WithdrawBalance

**Specification reference**

- see https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0020.md

**Other information and links**

### LINKED ISSUE #1256: implement FIP-0021: Correct quality calculation on expiration
**Task summary**

implement FIP-0021: Correct quality calculation on expiration

**Specification reference**

- see https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0021.md

**Other information and links**

### LINKED ISSUE #1257: implement FIP-0022: Bad deals don't fail PublishStorageDeals
**Task summary**

implement FIP-0022: Bad deals don't fail PublishStorageDeals

**Specification reference**

- see https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0022.md

**Other information and links**

### LINKED ISSUE #1258: implement FIP-0023: Break ties between tipsets of equal weight
**Task summary**

implement FIP-0023: Break ties between tipsets of equal weight

**Specification reference**

-  see https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0023.md

**Other information and links**

### LINKED ISSUE #1259: implement FIP-0024: BatchBalancer & BatchDiscount Post-HyperDrive Adjustment
**Task summary**

implement FIP-0024: BatchBalancer & BatchDiscount Post-HyperDrive Adjustment

**Specification reference**

-  see https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0024.md

**Other information and links**

### LINKED ISSUE #1318: Make lint uses Cargo clippy --all-features
**Describe the bug**

y'all use `cargo clippy --all-features`, which enables clippy against all possible features at once, in `make lint`. the bls-signatures dependency cannot have multiple features enabled at once - otherwise it redefines things and fails to compile, so make lint always spits out a ton of errors and fails. I'd recommend taking this flag out of `make lint` and also out of CI if it's there.

**To Reproduce**
1. run `make lint` against master

**Log output**

<details>
  <summary>Log Output</summary>
  
  ```Paste log output here
 error[E0428]: the name `key_gen` is defined multiple times
   --> /Users/laudiacay/.cargo/registry/src/github.com-1ecc6299db9ec823/bls-signatures-0.9.0/src/key.rs:248:1
    |
224 | fn key_gen<T: AsRef<[u8]>>(data: T) -> Fr {
    | ----------------------------------------- previous definition of the value `key_gen` here
...
248 | fn key_gen<T: AsRef<[u8]>>(data: T) -> Fr {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `key_gen` redefined here
    |
    = note: `key_gen` must be defined only once in the value namespace of this module

error[E0428]: the name `hash` is defined multiple times
  --> /Users/laudiacay/.cargo/registry/src/github.com-1ecc6299db9ec823/bls-signatures-0.9.0/src/signature.rs:82:1
   |
77 | pub fn hash(msg: &[u8]) -> G2 {
   | ----------------------------- previous definition of the value `hash` here
...
82 | pub fn hash(msg: &[u8]) -> G2 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `hash` redefined here
   |
   = note: `hash` must be defined only once in the value namespace of this module

error[E0428]: the name `verify_messages` is defined multiple times
   --> /Users/laudiacay/.cargo/registry/src/github.com-1ecc6299db9ec823/bls-signatures-0.9.0/src/signature.rs:300:1
    |
212 | / pub fn verify_messages(
213 | |     signature: &Signature,
214 | |     messages: &[&[u8]],
215 | |     public_keys: &[PublicKey],
...   |
223 | |     verify(signature, &hashes, public_keys)
224 | | }
    | |_- previous definition of the value `verify_messages` here
...
300 | / pub fn verify_messages(
301 | |     signature: &Signature,
302 | |     messages: &[&[u8]],
303 | |     public_keys: &[PublicKey],
...   |
344 | |     valid && pairing.finalverify(Some(&gtsig))
345 | | }
    | |_^ `verify_messages` redefined here
  ```
</details> 

**Expected behaviour**

This should work

**Screenshots**

**Environment (please complete the following information):**
 - OS: 
 - Rust version(e.g. `rustc --version`)
 - Branch/commit

**Other information and links**

### PR DESCRIPTION
**Pull request by noot**
**Summary of changes**
Changes introduced in this pull request:
- implements the v6 actors upgrade
- test vectors added as well

**Reference issue to close (if applicable)**

Closes #1255 
Closes #1256
Closes #1257
Closes #1259

also closes #1318

**Other information and links**

### CODE REVIEW

File: vm/actor/src/builtin/market/mod.rs [hunk 1]
```diff
@@ -188,9 +190,11 @@ impl Actor {
             recipient,
             METHOD_SEND,
             Serialized::default(),
-            amount_extracted,
+            amount_extracted.clone(),
         )?;
-        Ok(())
+        Ok(WithdrawBalanceReturn {
+            amount_withdrawn: amount_extracted,
+        })
```
Comment (noot) [inline vm/actor/src/builtin/market/mod.rs:197]: this fixes #1255

File: vm/actor/src/builtin/market/mod.rs [hunk 2]
```diff
@@ -248,9 +252,199 @@ impl Actor {
...
+                valid_deals.len(),
+                valid_deal_count
+            ));
+        }
+        if valid_deal_count <= 0 {
+            return Err(actor_error!(
+                ErrIllegalArgument,
+                "All deal proposals invalid"
+            ));
+        }
```
Comment (noot) [inline vm/actor/src/builtin/market/mod.rs:425]: I believe the code in this block is related to #1259

File: vm/actor/src/builtin/miner/mod.rs [hunk 3]
```diff
@@ -1227,6 +1236,19 @@ impl Actor {
...
+                // AggregateFee applied to fee debt to consolidate burn with outstanding debts
+                state.apply_penalty(&aggregate_fee)
+                .map_err(|e| {
+                    actor_error!(
+                        ErrIllegalState,
+                        "failed to apply penalty: {}",
+                        e
+                    )
+                })?;
+            }
```
Comment (noot) [inline vm/actor/src/builtin/miner/mod.rs:1251]: I believe this closes or is related to  #1259

File: vm/actor/src/builtin/miner/monies.rs [hunk 4]
```diff
@@ -281,17 +281,40 @@ pub fn locked_reward_from_reward(reward: TokenAmount) -> (TokenAmount, &'static
...
+    aggregate_size: i64,
+    gas_usage: &BigInt,
+    base_fee: &TokenAmount,
+) -> TokenAmount {
     let effective_gas_fee = max(base_fee, &*BATCH_BALANCER);
-    let network_fee_num = effective_gas_fee
-        * &*ESTIMATED_SINGLE_PROOF_GAS_USAGE
-        * aggregate_size
-        * &*BATCH_DISCOUNT_NUM;
+    let network_fee_num = effective_gas_fee * gas_usage * aggregate_size * &*BATCH_DISCOUNT_NUM;
```
Comment (noot) [inline vm/actor/src/builtin/miner/monies.rs:318]: related to  #1259

File: vm/actor/src/builtin/miner/monies.rs [hunk 5]
```diff
@@ -281,17 +281,40 @@ pub fn locked_reward_from_reward(reward: TokenAmount) -> (TokenAmount, &'static
 }
 
 lazy_static! {
-    static ref ESTIMATED_SINGLE_PROOF_GAS_USAGE: BigInt = BigInt::from(65733297);
+    static ref ESTIMATED_SINGLE_PROVE_COMMIT_GAS_USAGE: BigInt = BigInt::from(49299973);
+    static ref ESTIMATED_SINGLE_PRE_COMMIT_GAS_USAGE: BigInt = BigInt::from(16433324);
     static ref BATCH_DISCOUNT_NUM: BigInt = BigInt::from(1);
     static ref BATCH_DISCOUNT_DENOM: BigInt = BigInt::from(20);
-    static ref BATCH_BALANCER: BigInt = BigInt::from(2 * 1_000_000_000); // 2 * 1 nanoFIL
+    static ref BATCH_BALANCER: BigInt = BigInt::from(1_000_000_000) * BigInt::from(5); // 5 * 1 nanoFIL
```
Comment (olibearo) [inline vm/actor/src/builtin/miner/monies.rs:288]: @noot This also addresses #1259

**Pull request merged**